File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys
input=sys.stdin.readline
sys.setrecursionlimit(10**6)
from collections import defaultdict

def find(x):
    if x != p[x]:
        p[x] = find(p[x])
    return p[x]

def union(u,v):
    u = find(u)
    v = find(v)
    if u == v:
        return False
    if p[u] == p[v]:
        p[u] -= 1
    if p[u] < p[v]:
        p[v] = u
    else:
        p[u] = v
    return True

v,e = map(int,input().split())
p = [i for i in range(v+1)]
edge = []
for _ in range(e):
    a,b,c = map(int,input().split())
    edge.append((c,a,b))
edge.sort()

cnt = 0
ans = 0
for i in range(e):
    c,a,b = edge[i]
    if not union(a,b):
        continue
    ans += c
    cnt += 1
    if cnt == v-1:
        break

print(ans)

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.util.*;
import java.io.*;

class Main {
	static int[] parent;

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		Edge edges[] = new Edge[m];
		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine(), " ");
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			int v = Integer.parseInt(st.nextToken());
			edges[i] = new Edge(a, b, v);
		}
		parent = new int[n + 1];
		for (int i = 1; i <= n; i++) {
			parent[i] = i;
		}
		Arrays.sort(edges);
		int ans = 0;
		int useCnt = 0;
		int idx = 0;
		while (useCnt < n - 1) {
			Edge cur = edges[idx++];
			if (union(cur.a, cur.b)) {
				ans += cur.value;
				useCnt++;
			}
		}
		System.out.println(ans);
	}

	static class Edge implements Comparable<Edge> {
		int a, b, value;

		public Edge(int a, int b, int value) {
			this.a = a;
			this.b = b;
			this.value = value;
		}

		@Override
		public int compareTo(Edge o) {
			return Integer.compare(this.value, o.value);
		}
	}

	static int find(int x) {
		if (x == parent[x])
			return x;
		return parent[x] = find(parent[x]);
	}

	static boolean union(int x, int y) {
		x = find(x);
		y = find(y);
		if (x == y) {
			return false;
		} else {
			parent[y] = x;
			return true;
		}
	}
}